lti_consumer/lti_1p3/extensions/rest_framework/serializers.py

```diff
@@ -89,3 +90,159 @@ class Meta:
+      "comment" : "This is exceptional work.",
+      "activityProgress" : "Completed",
+      "gradingProgress": "FullyGraded",
+      "userId" : "5323497"
+    }
+
+    Reference:
+    https://www.imsglobal.org/spec/lti-ags/v2p0#example-application-vnd-ims-lis-v1-score-json-representation
+    """
+
+    # NOTE: `serializers.DateTimeField` always outputs the value in the local timezone of the server running the code
+    # This is because Django is time aware (see settings.USE_TZ) and because Django is unable to determine the timezone
+    # of the person making the API request, thus falling back on the local timezone. As such, since all outputs will
+    # necessarily be in a singular timezone, that timezone should be `utc`
+    timestamp = serializers.DateTimeField(input_formats=[ISO_8601], format=ISO_8601, default_timezone=timezone.utc)
+    scoreGiven = serializers.FloatField(source='score_given', required=False, allow_null=True, default=None)
+    scoreMaximum = serializers.FloatField(source='score_maximum', required=False, allow_null=True, default=None)
+    comment = serializers.CharField(required=False, allow_null=True)
+    activityProgress = serializers.CharField(source='activity_progress')
+    gradingProgress = serializers.CharField(source='grading_progress')
+    userId = serializers.CharField(source='user_id')
+
+    def validate_timestamp(self, value):
+        """
+        Ensure that if an existing record is being updated, that the timestamp is in the after the existing one
+        """
+        if self.instance:
+            if self.instance.timestamp > value:
+                raise serializers.ValidationError('Score timestamp can only be updated to a later point in time')
+
+            if self.instance.timestamp == value:
+                raise serializers.ValidationError('Score already exists for the provided timestamp')
+
+        return value
+
+    def is_valid(self, raise_exception=False):
+        """
+        Since most validation is currently done on the model, we want to make sure the current model will save
+        """
+
+        # If it isn't valid already, return those errors
+        if not super().is_valid(raise_exception=raise_exception):
+            return False
+
+        # `self.validated_data` is only set if `super().is_valid()` succeeds
+        new_instance = self.Meta.model(**self.validated_data)
+        try:
+            # Validate the instance
+            new_instance.full_clean()
+        except serializers.DjangoValidationError as exc:
+            errors = {key: value for key, value in exc.message_dict.items() if key != 'line_item'}
+            if errors:
+                # Unset any existing validated data because the model failed to validate
+                self._validated_data = {}  # pylint: disable=attribute-defined-outside-init
+
+                if raise_exception:
+                    raise serializers.ValidationError(errors)
+
+                return False
+
+        return True
```

Comment:
@pcockwell I think we can replace this by simply duplicating the validation in the serializer.
The main issue with this implementation is that it does a 2-step validation process, where if I send a request like this:
```
{"timestamp": "2020-09-27T18:54:36.740+00:00","scoreGiven":123, "comment" : "This is exceptional work.","activityProgress" : "Completed","gradingProgress": "FullyGraded"}
```
It'll first fail with `userId` not present, and after fixing it, it'll throw `maximumGrade` not present.

```suggestion
    def validate_scoreGiven(self, value):
        """
        Ensure that if an existing record is being updated, that the timestamp is in the after the existing one
        """
        if 'scoreMaximum' not in self.initial_data:
            raise serializers.ValidationError('ScoreMaximum is a required field when setting a scoreGiven value.')

        return value
```

I prefer simpler code, even if that means some duplication. It's easier for maintainability.